modules/transactionpool/update_test.go

```diff
@@ -40,3 +41,104 @@ func TestArbDataOnly(t *testing.T) {
 		t.Error("transaction was not cleared from the transaction pool")
 	}
 }
+
+// TestValidRevertedTransaction verifies that if a transaction appears in a
+// block's reverted transactions, it is added correctly to the pool.
+func TestValidRevertedTransaction(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	tpt, err := createTpoolTester(t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tpt.Close()
+
+	tpt2, err := blankTpoolTester(t.Name() + "-tpt2")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tpt2.Close()
+
+	// connect the testers and wait for them to have the same current block
+	err = tpt2.gateway.Connect(tpt.gateway.Address())
+	if err != nil {
+		t.Fatal(err)
+	}
+	success := false
+	for start := time.Now(); time.Since(start) < time.Minute; time.Sleep(time.Millisecond * 100) {
+		if tpt.cs.CurrentBlock().ID() == tpt2.cs.CurrentBlock().ID() {
+			success = true
+			break
+		}
+	}
+	if !success {
+		t.Fatal("testers did not have the same block height after one minute")
+	}
+
+	// disconnect the testers
+	err = tpt2.gateway.Disconnect(tpt.gateway.Address())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// make some transactions on tpt
+	var txnSets [][]types.Transaction
+	for i := 0; i < 5; i++ {
+		txns, err := tpt.wallet.SendSiacoins(types.SiacoinPrecision.Mul64(1000), types.UnlockHash{})
+		if err != nil {
+			t.Fatal(err)
+		}
+		txnSets = append(txnSets, txns)
+	}
+	// mine some blocks to cause a re-org
+	for i := 0; i < 3; i++ {
+		_, err = tpt.miner.AddBlock()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	// put tpt2 at a higher height
+	for i := 0; i < 10; i++ {
+		_, err = tpt2.miner.AddBlock()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// connect the testers and wait for them to have the same current block
+	err = tpt.gateway.Connect(tpt2.gateway.Address())
+	if err != nil {
+		t.Fatal(err)
+	}
+	success = false
+	for start := time.Now(); time.Since(start) < time.Minute; time.Sleep(time.Millisecond * 100) {
+		if tpt.cs.CurrentBlock().ID() == tpt2.cs.CurrentBlock().ID() {
+			success = true
+			break
+		}
+	}
+	if !success {
+		t.Fatal("testers did not have the same block height after one minute")
+	}
+
+	// verify the transaction pool still has the reorged txns
+	for _, txnSet := range txnSets {
+		for _, txn := range txnSet {
+			_, _, exists := tpt.tpool.Transaction(txn.ID())
```

Comment:
it's not going to, because you don't broadcast your existing tpool when someone joins the network.

I guess that's something we should consider, but it's out of scope for this PR either way.